crates/ty_python_semantic/src/types/signatures.rs

```diff
@@ -569,6 +571,63 @@ impl<'db> CallableSignature<'db> {
             }
         }
     }
+
+    /// Returns `true` if any signature in this callable has gradual parameters (`...`).
+    pub(crate) fn has_gradual_parameters(&self) -> bool {
+        self.overloads.iter().any(|sig| sig.parameters.is_gradual())
+    }
+
+    /// Materialize only the return types of all signatures, preserving parameters as-is.
+    ///
+    /// This is used when wrapping gradual callables in `Top[...]` or `Bottom[...]` - we want
+    /// to preserve the gradual parameters but materialize the return types (which are in
+    /// covariant position).
+    pub(crate) fn materialize_return_types(
+        &self,
+        db: &'db dyn Db,
+        materialization_kind: MaterializationKind,
+    ) -> Self {
+        Self::from_overloads(
+            self.overloads
+                .iter()
+                .map(|sig| sig.materialize_return_type(db, materialization_kind)),
+        )
+    }
+
+    /// Check whether the return types of this callable have the given relation to the return
+    /// types of another callable.
+    pub(crate) fn return_types_have_relation_to(
+        &self,
+        db: &'db dyn Db,
+        other: &Self,
+        inferable: InferableTypeVars<'_, 'db>,
+        relation: TypeRelation<'db>,
+        relation_visitor: &HasRelationToVisitor<'db>,
+        disjointness_visitor: &IsDisjointVisitor<'db>,
+    ) -> ConstraintSet<'db> {
+        // For each overload in self, the return type must have the relation to
+        // the return type of some overload in other.
+        self.overloads.iter().when_all(db, |self_sig| {
+            let Some(self_return_ty) = self_sig.return_ty else {
+                // No return type means Never, which is a subtype of everything
+                return ConstraintSet::from(true);
+            };
```

Comment:
This looks incorrect. If the function doesn't have an annotated return type, it returns `Unknown`, not `Never`. `Unknown` is:
- Only a subtype of `object`
- Redundant with other dynamic types, unions including dynamic types, and/or `object`
- Assignable to everything